### PR TITLE
Changes for Piers 

### PR DESCRIPTION
**Pull request by emveepee**
- incorporate changes from PR 266 Omega final.  
- reset to optional minimum inputstream adaptive.
- NextPVR has no group API calls.  Cache and compress channel list to avoid extra downloads.  Compress to save on disk for large IPTV channel lists with zlib potential future use with VOD 
- update to tinyxml2 version 10 used by Kodi

### CODE REVIEW

**Comment by emveepee** [comment]
@ksooo @phunkyfish could you please review.

**Comment by phunkyfish** [comment]
There is a merge commit that should not be there. Then two commits with the same name that doesn't make sense. Can you clean that up?

**Comment by emveepee** [comment]
> There is a merge commit that should not be there. Then two commits with the same name that doesn't make sense. Can you clean that up?

@phunkyfish yeah I couldn't figure out how to merge in those conflicted files that got changed twice after I started this.  Can you help me out on how to do that?

**Comment by phunkyfish** [comment]
I would just create a new branch and cherry pick the commits that you need. Then rename the branch and push it to the PR

**Comment by emveepee** [comment]
> I would just create a new branch and cherry pick the commits that you need. Then rename the branch and push it to the PR

Could I go back to my original approach noting I should have named commits differently,  Hopefully the following changes are clearer.

**Comment by ksooo** [comment]
The merge commit "[Update to PVR API 9](https://github.com/kodi-pvr/pvr.nextpvr/pull/269/commits/fd0e894d4adaa76728cf3bfb7c439c79e8e83100)" should be renoved from this PR.

File: src/BackendRequest.cpp
Comment (ksooo) [inline src/BackendRequest.cpp]: This looks like test code. Is it by intention to have it here?

File: src/BackendRequest.cpp
Comment (ksooo) [inline src/BackendRequest.cpp]: Remove, as not needed?

File: src/BackendRequest.h
Comment (ksooo) [inline src/BackendRequest.h]: `std::string` xml -> `const std::string&`

File: src/EPG.cpp
Comment (ksooo) [inline src/EPG.cpp]: Remove the `=`, it is not needed here.

File: src/pvrclient-nextpvr.cpp
Comment (ksooo) [inline src/pvrclient-nextpvr.cpp]: Remove empty line?

File: src/BackendRequest.cpp
Comment (emveepee) [reply src/BackendRequest.cpp]: Certainly was

File: src/EPG.cpp
Comment (emveepee) [reply src/EPG.cpp]: Yes that it is in Omega too (came from the port).  Fixed here not in Omega.

File: src/pvrclient-nextpvr.cpp
Comment (emveepee) [reply src/pvrclient-nextpvr.cpp]: Done

File: src/BackendRequest.cpp
Comment (emveepee) [reply src/BackendRequest.cpp]: Done

File: src/BackendRequest.h
Comment (emveepee) [reply src/BackendRequest.h]: Done

**Comment by emveepee** [comment]
> The merge commit "[Update to PVR API 9](https://github.com/kodi-pvr/pvr.nextpvr/pull/269/commits/fd0e894d4adaa76728cf3bfb7c439c79e8e83100)" should be renoved from this PR.

@phunkyfish @ksooo I redid all the commits and was able to rebase correctly to remove this.  I have one small change to make to change Channels.cpp back to the original filename on line 356   "/public/LiveStreams.xml";  and will do that after the changes are approved or if the review finds anything else.

File: pvr.nextpvr/changelog.txt
Comment (ksooo) [inline pvr.nextpvr/changelog.txt]: Aren't you explaining in line 4 to 6 what PR2666 was about? Then you could remove line 3, which says nothing to normal end users.

File: src/Recordings.cpp
Comment (ksooo) [inline src/Recordings.cpp]: That's actually not needed. Default ctor of PVRReocrding class will do that. Same for line 324 and 325. Not a bug, though. Just efficiency of code.

File: src/buffers/ClientTimeshift.h
Comment (ksooo) [inline src/buffers/ClientTimeshift.h]: std::time

File: src/EPG.cpp
Comment (ksooo) [inline src/EPG.cpp]: Remove space before and after the value?

File: src/EPG.cpp
Comment (ksooo) [inline src/EPG.cpp]: Could this be `static` to avoid construction on every function call?

File: src/EPG.cpp
Comment (ksooo) [inline src/EPG.cpp]: 16? Magic number anti pattern! Maybe introduce a constant so that one can see immediately what this value stands for. Same for the magic 4 and 0 in this line.

File: src/EPG.cpp
Comment (ksooo) [inline src/EPG.cpp]: std::atoi

File: src/Recordings.cpp
Comment (ksooo) [inline src/Recordings.cpp]: Could this be `static`?

File: src/Recordings.cpp
Comment (ksooo) [inline src/Recordings.cpp]: `static`?

File: src/Recordings.cpp
Comment (ksooo) [inline src/Recordings.cpp:530]: `static` std::regex instance?

File: src/EPG.cpp
Comment (ksooo) [inline src/EPG.cpp:150]: `static` std::regex instance?

File: pvr.nextpvr/changelog.txt
Comment (emveepee) [reply pvr.nextpvr/changelog.txt]: Fair enough remove comments about the port

File: src/buffers/ClientTimeshift.h
Comment (emveepee) [reply src/buffers/ClientTimeshift.h]: Done

File: src/EPG.cpp
Comment (emveepee) [reply src/EPG.cpp]: Sure.

File: src/EPG.cpp
Comment (emveepee) [reply src/EPG.cpp]: Done

File: src/Recordings.cpp
Comment (emveepee) [reply src/Recordings.cpp]: Sure

File: src/Recordings.cpp
Comment (emveepee) [reply src/Recordings.cpp:530]: Yes

File: src/EPG.cpp
Comment (emveepee) [reply src/EPG.cpp:150]: yes

File: src/Recordings.cpp
Comment (emveepee) [reply src/Recordings.cpp]: OK

File: src/EPG.cpp
Comment (emveepee) [reply src/EPG.cpp]: OK use system defines and comment on the 4 instead of constant

File: src/EPG.cpp
Comment (emveepee) [reply src/EPG.cpp]: Done

File: src/Recordings.cpp
Comment (emveepee) [reply src/Recordings.cpp]: Removed all three

File: pvr.nextpvr/addon.xml.in
Comment (phunkyfish) [inline pvr.nextpvr/addon.xml.in]: This looks wrong. For Piers this would need to be starting with `22.`

**Review by phunkyfish — CHANGES_REQUESTED** [review]
I see all the comments have been addressed. Just one thing to change.

File: pvr.nextpvr/addon.xml.in
Comment (emveepee) [reply pvr.nextpvr/addon.xml.in]: Historically I have left it like this because it means a don't have to update it with each minimum version of inputstream.ffmpegdirect,  even 22.0.0 will be wrong over the life of Piers, the bigger problem occurs when optional="true" is removed as happened with Piers.  Also the minimum does not give more versions to install from the addon, only API compatible versions are presented.  

The use case (client requested transcoding) is rare enough that I will be willing to support it if someone actually tries to install everything manually based only on addon.xml

Comment (ksooo) [reply pvr.nextpvr/addon.xml.in]: > even 22.0.0 will be wrong over the life of Piers

Not necessarily, only if inputstream.ffmpegdirect changes incompatible, which should only happen during Piers development phase not after Kodi 22.0 was officially released.

Comment (ksooo) [reply pvr.nextpvr/addon.xml.in]: Everything < 22.0.0 will not work with Piers. If you keep it like this you can also remove whatever you are doing with this addon in pvr.nextpvr. That would have the same effect. :-)

Comment (emveepee) [reply pvr.nextpvr/addon.xml.in]: Right but the installer system looks at the repo mirror and it shows only Piers builds and it figured out 22.0.0 for me when I removed the minversion completely.

image

Comment (phunkyfish) [reply pvr.nextpvr/addon.xml.in]: minversion was very important in the early days off ffmpegdirect development. Now, not so much. 

Is your addon’s use of ffmpegdirect truly optional? I.e. you use the addon API to figure out what’s installed and only use ffmpegdirect if it’s available?

Comment (emveepee) [reply pvr.nextpvr/addon.xml.in]: > Is your addon’s use of ffmpegdirect truly optional? I.e. you use the addon API to figure out what’s installed and only use ffmpegdirect if it’s available?

Normally playback is direct play but there is an option to support server side transcoding and in that mode there is another option to support timeshifted transcoded playback with ffmpegdirect.  And yes there are warnings if this use is attempted when it isn't installed or enabled.  https://github.com/kodi-pvr/pvr.nextpvr/blob/fec5c553ca5bc11d1dff22132824715b920577da/src/pvrclient-nextpvr.cpp#L276

Comment (emveepee) [reply pvr.nextpvr/addon.xml.in]: > Everything < 22.0.0 will not work with Piers. If you keep it like this you can also remove whatever you are doing with this addon in pvr.nextpvr. That would have the same effect. :-)

@ksooo I do not really understanding your comment since unsupported versions won't be in the repo https://mirrors.kodi.tv/addons/piers/addons.xml.gz  I don't think I have to remove any code.  

My point about 22.0.0 was simply that version doesn't exist so it is just as invalid as the lower version I was using, I think removing the minversion is the way to go.

Comment (phunkyfish) [reply pvr.nextpvr/addon.xml.in]: Let’s try removing minversion in nextpvr and see how it goes.

**Review by phunkyfish — APPROVED** [review]

**Pull request merged**
